### PR TITLE
start unit tests

### LINKED ISSUE #86: start unit tests
This is a GUI so unit testing might be a challenge.  Still, some parts can be tested without GUI.

### PR DESCRIPTION
**Pull request by prjemian**
- close #86

### CODE REVIEW

**Comment by prjemian** [comment]
@rodolakis Ready for review.  So far, these tests do not require an active _tiled_ server.

File: gemviz/tests/test_init.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+import pytest
+
+import gemviz  # TODO: Can this become a relative import?
```
Comment (prjemian) [inline gemviz/tests/test_init.py:3]: The problem is how to import symbols from `../__init__.py`

File: gemviz/tests/test_init.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+@pytest.mark.parametrize(
+    "var_name, where, fragment",
+    [
+        ["LICENSE_FILE", "end", "LICENSE"],
+        ["UI_DIR", "end", "resources"],
+    ],
+)
+def test_paths(var_name, where, fragment):
+    apply_str_assertion(str(getvar(var_name)), where, fragment)
```
Comment (prjemian) [inline gemviz/tests/test_init.py:55]: Can these `test_paths()` be combined with previous by applying `str(getvar(var_name))` to all?

File: gemviz/tests/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
+import datetime
+
+import pytest
+from PyQt5 import QtWidgets
+
+from .. import utils
+
+# TODO: run_in_thread
```
Comment (prjemian) [inline gemviz/tests/test_utils.py:8]: Copy the test from apstools?

File: gemviz/tests/test_aboutdialog.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from PyQt5 import QtWidgets
+
+from .. import aboutdialog
+
+
+def test_about_starts(qtbot):
```
Comment (prjemian) [inline gemviz/tests/test_aboutdialog.py:6]: Let's add a `requirements-dev.txt` file to describe packages needed to develop this code.  A separate file will be used to describe requirements to build the documentation.

File: gemviz/tests/test_aboutdialog.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    """About dialog should start with no errors."""
+
+    class SetStatusWidget(QtWidgets.QWidget):
+        """Test class with setStatus() method."""
+
+        def setStatus(self, status):
+            pass
+
+    widget_with_status = SetStatusWidget()
```
Comment (prjemian) [inline gemviz/tests/test_aboutdialog.py:15]: call it `fake_main_window`

File: gemviz/tests/test_chartview.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    if len(curves) > 0:
+        curve = curves[0].curve
+        assert curve.yData.shape == y.shape  # compare shapes
+        assert (curve.yData == y).all()  # compare values
+        if x is None:
+            assert curve.xData.shape == y.shape
+            assert (curve.xData != x).all()
+        else:
+            assert curve.xData.shape == x.shape
+            assert (curve.xData == x).all()
```
Comment (prjemian) [inline gemviz/tests/test_chartview.py:84]: ... de-indent this line

File: gemviz/tests/test_utils.py (hunk 3 above)
Comment (prjemian) [reply gemviz/tests/test_utils.py:8]: It seems there is no test code now in apstools for `run_in_thread()`.  However, [there is code ](https://github.com/search?q=repo%3ABCDA-APS%2Fapstools%20run_in_thread&type=code) that uses this function.

So, no changes now.

File: gemviz/tests/test_aboutdialog.py (hunk 4 above)
Comment (prjemian) [reply gemviz/tests/test_aboutdialog.py:6]: Partly, content is documented in the (new) `testing.md` file.

**Comment by prjemian** [comment]
When building documentation locally, notice this message from `sphinx-build`:

```
WARNING: html_static_path entry '_static' does not exist
```

Since git does not include empty directories, we need to create this one and put a file there, at least until we have some real content (such as screen views).

File: gemviz/tests/test_aboutdialog.py (hunk 4 above)
Comment (prjemian) [reply gemviz/tests/test_aboutdialog.py:6]: Converted `.md` to .rst` and added to docs

File: gemviz/tests/test_chartview.py (hunk 6 above)
Comment (prjemian) [reply gemviz/tests/test_chartview.py:84]: No.  The comparisons (`!=` and `==`) are also different.

File: gemviz/tests/test_chartview.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    pw = chart.plot_widget
+    curves = pw.getPlotItem().curves
+
+    if len(curves) > 0:
+        curve = curves[0].curve
+        assert curve.yData.shape == y.shape  # compare shapes
+        assert (curve.yData == y).all()  # compare values
+        if x is None:
+            assert curve.xData.shape == y.shape
+            assert (curve.xData != x).all()
```
[the remark this answers is not in the document]
Comment (prjemian) [reply gemviz/tests/test_chartview.py:81]: no, see below

**Comment by prjemian** [comment]
@rodolakis : Ready for review.  Includes (and adds to) the `doc/` directory added by #176.

**Review by rodolakis — APPROVED** [review]
Sounds good.

**Pull request merged**